client/src/api/fetcher.ts

```diff
@@ -0,0 +1,120 @@
+/* eslint-disable @typescript-eslint/no-throw-literal */
+/* eslint-disable @typescript-eslint/ban-types */
+import type { SimpleError } from './responses';
+
+const resolveUrl = (
+  url: string,
+  _queryParams: Record<string, unknown> = {},
+  pathParams: Record<string, string> = {}
+) => {
+  return url.replace(/\{\w*\}/g, (key) => pathParams[key.slice(1, -1)]);
+};
+
+// Typed only the subset of the spec we actually use (to be able to build a simple mock)
+export type FetchImpl = (
+  url: string,
+  init?: { body?: string; headers?: Record<string, string>; method?: string }
+) => Promise<{ ok: boolean; status: number; json(): Promise<any> }>;
+
+export type FetcherExtraProps = {
+  apiUrl: string;
+  workspacesApiUrl: string;
+  fetchImpl: FetchImpl;
+  apiKey: string;
+};
+
+export type FetcherOptions<TBody, THeaders, TQueryParams, TPathParams> = {
+  url: string;
+  method: string;
+  body?: TBody;
+  headers?: THeaders;
+  queryParams?: TQueryParams;
+  pathParams?: TPathParams;
+};
+
+const fallbackError: SimpleError = { message: 'Network response was not ok' };
+
+function baseURLForWorkspace(workspacesApiUrl: string, workspace: string) {
+  if (!workspacesApiUrl.includes('{workspaceId}')) return workspacesApiUrl;
```

Comment:
I tried to take into account the scenario where they have CNAME links of their workspace url, but in the docs we will need to say that ``{workspaceId}`` is a restricted keyword that injects the workspace id.